### PR TITLE
ServiceAccounts: add role picker to service accounts list

### PR DESCRIPTION
**Pull request by Jguer**
**What this PR does / why we need it**:

- Add role picker to service accounts list

AC disabled:

image

AC enabled:

image

---

What is not implemented:

- Proper action in AC metadata for updating service account role (missing org.users) - to be discussed with rest of AC team

Closes https://github.com/grafana/grafana-enterprise/issues/1977

Closes https://github.com/grafana/grafana-enterprise/issues/2618

### CODE REVIEW

File: public/app/features/serviceaccounts/state/actions.ts [hunk 1]
```diff
@@ -36,18 +37,17 @@ export function loadServiceAccountTokens(saID: number): ThunkResult<void> {
...
     }
   };
 }
 
-export function updateServiceAccount(saID: number): ThunkResult<void> {
+export function updateServiceAccount(serviceAccount: ServiceAccountDTO): ThunkResult<void> {
   return async (dispatch) => {
-    // TODO: implement on backend
-    await getBackendSrv().patch(`${BASE_URL}/${saID}`, {});
+    await getBackendSrv().patch(`/api/org/users/${serviceAccount.id}`, { role: serviceAccount.role });
```
Comment (Jguer) [inline public/app/features/serviceaccounts/state/actions.ts:50]: :information_source: we should make an update route for service accounts and block off SA editing in users endpoint

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx [hunk 2]
```diff
@@ -33,12 +34,46 @@ const mapDispatchToProps = {
...
+  navModel,
+  serviceAccounts,
+  isLoading,
+}) => {
   const styles = useStyles2(getStyles);
+  const [roleOptions, setRoleOptions] = useState<Role[]>([]);
+  const [builtinRoles, setBuiltinRoles] = useState<Record<string, Role[]>>({});
 
   useEffect(() => {
+    async function fetchOptions() {
```
Comment (Clarity-89) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:49]: I think the same logic is used in other places where role picker is used, I think it'd make sense to extract it into a custom hook.

Comment (Jguer) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:49]: Yes, this one particularly comes from team list

Comment (Jguer) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx]: It doesn't really map to exactly the same usage as in other places (and looking at it I'm pretty sure it's bound to change soon) but I've moved this logic to redux

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx [hunk 3]
```diff
@@ -4,41 +4,62 @@ import { Icon, LinkButton, useStyles2 } from '@grafana/ui';
...
   loadServiceAccounts,
+  fetchACOptions,
   updateServiceAccount,
   removeServiceAccount,
 };
 
 const connector = connect(mapStateToProps, mapDispatchToProps);
 
-const ServiceAccountsListPage: React.FC<Props> = ({ loadServiceAccounts, navModel, serviceAccounts, isLoading }) => {
+const ServiceAccountsListPage: React.FC<Props> = ({
```
Comment (eleijonmarck) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:39]: love that we have FC! ❤️ 


it's so hot right now!

Comment (Clarity-89) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:39]: We are moving away from React.FC in our code and it shouldn't be used for new components. The change is quite simple though:
```
const ServiceAccountsListPage = ({
  loadServiceAccounts,
  fetchACOptions,
  updateServiceAccount,
  navModel,
  serviceAccounts,
  isLoading,
  roleOptions,
  builtInRoles,
}: Props)

**Review by eleijonmarck — APPROVED** [review]
Through review and nothing really found. Just a good nice read! well done

File: public/app/features/serviceaccounts/state/actions.ts [hunk 4]
```diff
@@ -1,9 +1,30 @@
...
+import { fetchBuiltinRoles, fetchRoleOptions } from 'app/core/components/RolePicker/api';
 
 const BASE_URL = `/api/serviceaccounts`;
 
+export function fetchACOptions(): ThunkResult<void> {
+  return async (dispatch) => {
+    try {
+      const options = await fetchRoleOptions();
+      const builtInRoles = await fetchBuiltinRoles();
+      dispatch(acOptionsLoaded(options));
```
Comment (Clarity-89) [inline public/app/features/serviceaccounts/state/actions.ts:20]: Not sure if it makes any difference here but maybe we'd dispatch the options loaded already after they were fetch? Currently we're also waiting for the roles to be fetched.

**Review by Clarity-89 — APPROVED** [review]
Looks good, left a few minor comments.

File: public/app/features/serviceaccounts/state/actions.ts (hunk 4 above)
Comment (Jguer) [reply public/app/features/serviceaccounts/state/actions.ts:20]: It should make any difference in this case,  but I'll change it so it "makes sense"

**Pull request merged**